### ISSUE TITLE
advanceTo skips first occurence when using bysetpos

### ISSUE THREAD

**Issue by gaspardpetit** [mangstadt/biweekly#89]
When using `.advanceTo` on a DateIterator that iterates over a RRULE with a BYSETPOS rule, the immediate next occurrence is skipped:

```{java}
@Test
public void givenRRULE_whenAdvanceTo_ensureNextIsNextOccurrence()
{
	// given
	final String rrule = "FREQ=MONTHLY;BYSETPOS=3;BYDAY=MO"; // every 3rd Monday of the month
	final Date startDate = Convert.toDate("2019-01-01T00:00:00Z"); // starting in 2019

	// when
	RecurrenceRuleScribe scribe = new RecurrenceRuleScribe();
	ParseContext context = new ParseContext();
	context.setVersion(ICalVersion.V2_0);
	RecurrenceRule recurrenceRule = scribe.parseText(rrule, null, new ICalParameters(), context);
	DateIterator iter = recurrenceRule.getDateIterator(Convert.toDate(startDate), TimeZone.getTimeZone("UTC"));

	Date evalDate = Convert.toDate("2019-08-01T00:00:00Z");
	iter.advanceTo(evalDate);
	Instant nextOccurrence = iter.next().toInstant();
	
	// ensure
	Assertions.assertEquals(Convert.toDate("2019-08-19T00:00:00Z").toInstant(), nextOccurrence);
}
```

The code aboves asserts with this message:

```
Expected :2019-08-19T00:00:00Z
Actual   :2019-09-16T00:00:00Z
```

As you can see, the RRULE identifies the 3rd monday of each month as a valid date. When skipping to the beginning of August, the iterator skips the 3rd Monday of August (August 19) and goes straight to the 3rd Monday of September.

This problem makes it difficult to skip to the current date and evaluate what the next occurrence of a RRULE is.

**Comment by mangstadt** [comment]
Yes, I get the same result when I run your code.

The iterator appears to be calculating the dates correctly, as shown in the code sample below. The problem seems to be with the `advanceTo()` method not knowing the correct place to stop.

The code below outputs the 3rd Monday of every month in the year 2019:

```java
TimeZone utc = TimeZone.getTimeZone("UTC");

DateFormat df = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
df.setTimeZone(utc);

Recurrence recurrence = new Recurrence.Builder(Frequency.MONTHLY)
  .bySetPos(3)
  .byDay(DayOfWeek.MONDAY)
.build();

Date startDate = df.parse("2019-01-01 00:00:00");
DateIterator iter = recurrence.getDateIterator(startDate, utc);

for (int i = 0; i < 12; i++) {
  Date next = iter.next();
  System.out.println(df.format(next));
}
```

Output:

```
2019-01-21 00:00:00
2019-02-18 00:00:00
2019-03-18 00:00:00
2019-04-15 00:00:00
2019-05-20 00:00:00
2019-06-17 00:00:00
2019-07-15 00:00:00
2019-08-19 00:00:00
2019-09-16 00:00:00
2019-10-21 00:00:00
2019-11-18 00:00:00
2019-12-16 00:00:00
```

**Comment by mangstadt** [comment]
Fixed in 4cbfb3b320cdb387d4c91c4e22815284b0699662.

The `advanceTo` method contains code that jumps the iterator ahead to improve performance. I disabled this optimization for recurrences that contain BYSETPOS values.

This is bug exists in the original google-rfc-2245 code as well.
